### PR TITLE
lightning: add complex integration tests for lightning post-import conflict detection "replace" mode

### PR DESCRIPTION
**Pull request by lyzx2001**
### What problem does this PR solve?

Issue Number: ref #45774 

Problem Summary:

Currently lightning only supports "remove" mode for post-import conflict detection, but many customers request lightning to support "replace" mode for lightning post-import conflict detection.

We would like to support "replace" mode for lightning post-import conflict detection:
To resolve rows with conflict, instead of deleting all the rows that are engaged in conflict (the algorithm for remove), we delete some of the rows with conflict and reserve other rows that can be kept and not cause conflict anymore. Under this circumstance, we only delete the necessary rows to resolve conflicts, so that we can keep more original rows than remove mode as long as the conflicts are resolved.

The algorithms for index KV checking is contained in https://github.com/pingcap/tidb/pull/45926

The algorithms for data KV checking is contained in https://github.com/pingcap/tidb/pull/46763

This PR contains complex unit tests and integration tests to test the correctness and performance of the algorithms.

The tests vary in the setting of engine batch-size to cover various complex cases .

### What is changed and how it works?

Demo code for 'replace' mode of lightning post-import conflict detection:

https://github.com/lyzx2001/tidb-conflict-replace

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  >   

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/lightning/errormanager/errormanager.go [hunk 1]
```diff
@@ -122,19 +124,19 @@ const (
 
 	insertIntoConflictErrorData = `
 		INSERT INTO %s.` + ConflictErrorTableName + `
```
Comment (lance6716) [inline br/pkg/lightning/errormanager/errormanager.go:126]: When changing the table structure, need to change `ConflictErrorTableName` as well.

File: br/pkg/lightning/errormanager/errormanager.go [hunk 2]
```diff
@@ -644,6 +662,26 @@ func (em *ErrorManager) ReplaceConflictKeys(
 				// Only if there is a->1 we dare to delete data KV with key "1".
 
 				if bytes.Equal(kvPair.Key, rawKey) && bytes.Equal(kvPair.Val, rawValue) {
+					if err := exec.Transact(ctx, "insert data conflict error record for conflict detection 'replace' mode",
```
Comment (lance6716) [inline br/pkg/lightning/errormanager/errormanager.go:665]: move the creating of `exec` closer to its usage.

File: br/pkg/lightning/errormanager/errormanager.go [hunk 3]
```diff
@@ -644,6 +662,26 @@ func (em *ErrorManager) ReplaceConflictKeys(
...
+						func(c context.Context, txn *sql.Tx) error {
+							sb := &strings.Builder{}
+							fmt.Fprintf(sb, insertIntoConflictErrorData, em.schemaEscaped)
+							var sqlArgs []interface{}
+							sb.WriteString(sqlValuesConflictErrorData)
+							sqlArgs = append(sqlArgs,
+								em.taskID,
+								tableName,
+								"null",
```
Comment (lance6716) [inline br/pkg/lightning/errormanager/errormanager.go:674]: maybe `nil` is OK to represent SQL NULL

Comment (lyzx2001) [reply br/pkg/lightning/errormanager/errormanager.go]: modified

**Review by lance6716 — APPROVED** [review]

File: br/pkg/lightning/errormanager/errormanager.go [hunk 4]
```diff
@@ -94,6 +95,7 @@ const (
 			raw_value   mediumblob NOT NULL COMMENT 'the value of the conflicted key',
 			raw_handle  mediumblob NOT NULL COMMENT 'the data handle derived from the conflicted key or value',
 			raw_row     mediumblob NOT NULL COMMENT 'the data retrieved from the handle',
+			is_data_kv  tinyint(1) NOT NULL,
```
Comment (D3Hunter) [inline br/pkg/lightning/errormanager/errormanager.go:98]: maybe `kv_group`, values can be `data`/`index`, or if we want finer grain, we can use index id as the name.

current is lgtm too

File: br/tests/lightning_duplicate_resolution_replace_multiple_keys_clustered_pk/data/dup_resolve.a-schema.sql [hunk 5]
```diff
@@ -0,0 +1,8 @@
+create table a (
+    a int primary key clustered,
+    b int not null,
+    c int not null,
+    d text,
+    key key_b(b),
+    key key_c(b)
```
Comment (D3Hunter) [inline br/tests/lightning_duplicate_resolution_replace_multiple_keys_clustered_pk/data/dup_resolve.a-schema.sql:7]: index on `c`?

File: br/tests/lightning_duplicate_resolution_replace_multiple_keys_clustered_pk/run.sh [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+run_sql 'admin check table dup_resolve.a'
+
+run_sql 'select count(*) from dup_resolve.a'
+check_contains 'count(*): 5'
+
+run_sql 'select * from dup_resolve.a'
+check_contains 'a: 1'
+check_contains 'b: 1'
+check_contains 'c: 1'
+check_contains 'd: 1.csv'
```
Comment (D3Hunter) [inline br/tests/lightning_duplicate_resolution_replace_multiple_keys_clustered_pk/run.sh:33]: more like `ignore` semantic in this case

File: br/tests/lightning_duplicate_resolution_replace_multiple_keys_nonclustered_pk/data/dup_resolve.a-schema.sql [hunk 7]
```diff
@@ -0,0 +1,8 @@
+create table a (
+    a int primary key nonclustered,
+    b int not null,
+    c int not null,
+    d text,
+    key key_b(b),
+    key key_c(b)
```
Comment (D3Hunter) [inline br/tests/lightning_duplicate_resolution_replace_multiple_keys_nonclustered_pk/data/dup_resolve.a-schema.sql:7]: doitto

**Review by D3Hunter — COMMENTED** [review]
rest lgtm

File: br/pkg/lightning/errormanager/errormanager.go (hunk 4 above)
Comment (lyzx2001) [reply br/pkg/lightning/errormanager/errormanager.go:98]: maybe current `is_data_kv` is simpler in implemention, it seems like it is not necessary to determine whether the value is `data` or `index` based on the result `tablecodec.IsRecordKey(conflictInfo.RawKey)`

File: br/tests/lightning_duplicate_resolution_replace_multiple_keys_clustered_pk/data/dup_resolve.a-schema.sql (hunk 5 above)
Comment (lyzx2001) [reply br/tests/lightning_duplicate_resolution_replace_multiple_keys_clustered_pk/data/dup_resolve.a-schema.sql]: modified

File: br/tests/lightning_duplicate_resolution_replace_multiple_keys_clustered_pk/run.sh (hunk 6 above)
Comment (lyzx2001) [reply br/tests/lightning_duplicate_resolution_replace_multiple_keys_clustered_pk/run.sh:33]: I did not quite understand. Here `1,1,1,1.csv` and `1,1,2,2.csv` have conflicts due to PK, and here in `replace` semantic we reserve the first row and delete the second row.

File: go.mod [hunk 8]
```diff
@@ -186,7 +186,7 @@ require (
 	github.com/eapache/go-resiliency v1.2.0 // indirect
 	github.com/eapache/go-xerial-snappy v0.0.0-20180814174437-776d5712da21 // indirect
 	github.com/eapache/queue v1.1.0 // indirect
-	github.com/fatih/structtag v1.2.0 // indirect
+	github.com/fatih/structtag v1.2.0
```
Comment (wuhuizuo) [inline go.mod:189]: Does it become a direct dependency go module?

Comment (lyzx2001) [reply go.mod:189]: This deletion was auto performed by `make tidy`

**Comment by lyzx2001** [comment]
> @wuhuizuo: **Potential Problems:**
> 
> 1. In the `config.go` file, the comments for `DupeResAlgRecord` and `DupeResAlgRemove` still mention the old `lightning_task_info.conflict_error_v1` table instead of the new `lightning_task_info.conflict_error_v2` table.
> 
> **Suggestions:**
> 
> 1. Update the comments for `DupeResAlgRecord` and `DupeResAlgRemove` to mention the new `lightning_task_info.conflict_error_v2` table instead of the old `lightning_task_info.conflict_error_v1` table.
> 
> ```diff
> diff --git a/br/pkg/lightning/config/config.go b/br/pkg/lightning/config/config.go
> index 6d71f35795b923..03d97522a4ebd2 100644
> --- a/br/pkg/lightning/config/config.go
> +++ b/br/pkg/lightning/config/config.go
> @@ -588,16 +588,16 @@ const (
>  	// DupeResAlgNone doesn't detect duplicate.
>  	DupeResAlgNone DuplicateResolutionAlgorithm = iota
>  
> -	// DupeResAlgRecord only records duplicate records to `lightning_task_info.conflict_error_v1` table on the target TiDB.
> +	// DupeResAlgRecord only records duplicate records to `lightning_task_info.conflict_error_v2` table on the target TiDB.
>  	DupeResAlgRecord
>  
>  	// DupeResAlgRemove records all duplicate records like the 'record' algorithm and remove all information related to the
> -	// duplicated rows. Users need to analyze the lightning_task_info.conflict_error_v1 table to add back the correct rows.
> +	// duplicated rows. Users need to analyze the lightning_task_info.conflict_error_v2 table to add back the correct rows.
>  	DupeResAlgRemove
>  
>  	// DupeResAlgReplace records all duplicate records like the 'record' algorithm, and remove some rows with conflict
>  	// and reserve other rows that can be kept and not cause conflict anymore. Users need to analyze the
> -	// lightning_task_info.conflict_error_v1 table to check whether the reserved data cater to their need and check whether
> +	// lightning_task_info.conflict_error_v2 table to check whether the reserved data cater to their need and check whether
>  	// they need to add back the correct rows
> 
> ......
> ```
> 
>  > Response is trunked for length limits.

In the code all `lightning_task_info.conflict_error_v1` have already been changed to `lightning_task_info.conflict_error_v2`, we may ignore this suggestion.

File: br/tests/lightning_duplicate_resolution_replace_multiple_keys_clustered_pk/run.sh (hunk 6 above)
Comment (D3Hunter) [reply br/tests/lightning_duplicate_resolution_replace_multiple_keys_clustered_pk/run.sh:33]: `ignore` will keep the first row met, like what's done here, `replace` will keep the last

Comment (lyzx2001) [reply br/tests/lightning_duplicate_resolution_replace_multiple_keys_clustered_pk/run.sh:33]: Here `batch-size = 1`, the actual import order is not the same as that presented in `dup_resolve.a.1`

**Comment by wuhuizuo** [comment]
> > @wuhuizuo: **Potential Problems:**
> > 
> > 1. In the `config.go` file, the comments for `DupeResAlgRecord` and `DupeResAlgRemove` still mention the old `lightning_task_info.conflict_error_v1` table instead of the new `lightning_task_info.conflict_error_v2` table.
> > 
> > **Suggestions:**
> > 
> > 1. Update the comments for `DupeResAlgRecord` and `DupeResAlgRemove` to mention the new `lightning_task_info.conflict_error_v2` table instead of the old `lightning_task_info.conflict_error_v1` table.
> > 
> > ```diff
> > diff --git a/br/pkg/lightning/config/config.go b/br/pkg/lightning/config/config.go
> > index 6d71f35795b923..03d97522a4ebd2 100644
> > --- a/br/pkg/lightning/config/config.go
> > +++ b/br/pkg/lightning/config/config.go
> > @@ -588,16 +588,16 @@ const (
> >  	// DupeResAlgNone doesn't detect duplicate.
> >  	DupeResAlgNone DuplicateResolutionAlgorithm = iota
> >  
> > -	// DupeResAlgRecord only records duplicate records to `lightning_task_info.conflict_error_v1` table on the target TiDB.
> > +	// DupeResAlgRecord only records duplicate records to `lightning_task_info.conflict_error_v2` table on the target TiDB.
> >  	DupeResAlgRecord
> >  
> >  	// DupeResAlgRemove records all duplicate records like the 'record' algorithm and remove all information related to the
> > -	// duplicated rows. Users need to analyze the lightning_task_info.conflict_error_v1 table to add back the correct rows.
> > +	// duplicated rows. Users need to analyze the lightning_task_info.conflict_error_v2 table to add back the correct rows.
> >  	DupeResAlgRemove
> >  
> >  	// DupeResAlgReplace records all duplicate records like the 'record' algorithm, and remove some rows with conflict
> >  	// and reserve other rows that can be kept and not cause conflict anymore. Users need to analyze the
> > -	// lightning_task_info.conflict_error_v1 table to check whether the reserved data cater to their need and check whether
> > +	// lightning_task_info.conflict_error_v2 table to check whether the reserved data cater to their need and check whether
> >  	// they need to add back the correct rows
> > 
> > ......
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > > Response is trunked for length limits.
> 
> In the code all `lightning_task_info.conflict_error_v1` have already been changed to `lightning_task_info.conflict_error_v2`, we may ignore this suggestion.

Yes, the AI misjudge it.

**Review by 3pointer — APPROVED** [review]

**Pull request merged**